starfish/display/stack.py

```diff
@@ -0,0 +1,222 @@
+import warnings
+from collections import OrderedDict
+from typing import Iterable, List, Optional, Set, Tuple, Union
+
+import numpy as np
+
+from starfish.imagestack.imagestack import ImageStack
+from starfish.intensity_table.intensity_table import IntensityTable
+from starfish.types import Features, Indices
+
+
+def _normalize_indices(indices: Iterable[Union[Indices, str]]) -> List[str]:
+    """convert all indices to strings"""
+
+    def _normalize(indices):
+        if isinstance(indices, str):
+            return indices
+        elif isinstance(indices, Indices):
+            return indices.value
+        else:
+            raise TypeError(f'passed indices must be str or Indices objects not {type(indices)}')
+
+    return list(_normalize(i) for i in indices)
+
+
+def _max_intensity_table_maintain_dims(
+    intensity_table: IntensityTable, dimensions: Set[Indices],
+) -> IntensityTable:
+    """
+    Maximum project an IntensityTable over dimensions, retaining singletons in place of dimensions
+    reduced by the maximum operation.
+
+    For example, xarray.max(Indices.CH.value) would usually produce a 2-d (features, rounds) output.
+    This function ensures that the output is instead (features, 1, rounds), by inserting singleton
+    dimensions in place of any projected axes.
+
+    Parameters
+    ----------
+    intensity_table : IntensityTable
+        Input intensities
+    dimensions : Set[Indices]
+        Dimensions to project
+
+    Returns
+    -------
+    IntensityTable :
+        full-dimensionality (3-d) IntensityTable
+
+    """
+    str_dimensions = _normalize_indices(dimensions)
+    initial_dimensions = OrderedDict(intensity_table.sizes)
+    projected_intensities = intensity_table.max(str_dimensions)
+    expanded_intensities = projected_intensities.expand_dims(str_dimensions)
+    return expanded_intensities.transpose(*tuple(initial_dimensions.keys()))
+
+
+def _mask_low_intensity_spots(
+    intensity_table: IntensityTable,
+    intensity_threshold: float
+) -> np.ndarray:
+    """
+    returns a flattened boolean mask (c order flattening) that is True where
```

Comment:
this flattening concept is foreign to me.  I thought it was some reduction operation whereas it is a reshaping operation.